zokrates_core/src/absy/from_ast.rs

```diff
@@ -1,66 +1,70 @@
-                .map(absy::SymbolDeclarationNode::from)
-                .chain(
-                    prog.constants
-                        .into_iter()
-                        .map(absy::SymbolDeclarationNode::from),
-                )
-                .chain(
-                    prog.functions
-                        .into_iter()
-                        .map(absy::SymbolDeclarationNode::from),
-                ),
-        )
-        .imports(
-            prog.imports
-                .into_iter()
-                .map(absy::ImportDirective::from)
-                .flatten(),
-        )
+    fn from(file: pest::File<'ast>) -> absy::Module<'ast> {
+        absy::Module::with_symbols(file.declarations.into_iter().flat_map(|d| match d {
+            pest::SymbolDeclaration::Import(i) => import_directive_to_symbol_vec(i),
+            pest::SymbolDeclaration::Constant(c) => vec![c.into()],
+            pest::SymbolDeclaration::Struct(s) => vec![s.into()],
+            pest::SymbolDeclaration::Function(f) => vec![f.into()],
+        }))
     }
 }
 
-impl<'ast> From<pest::ImportDirective<'ast>> for absy::ImportDirective<'ast> {
-    fn from(import: pest::ImportDirective<'ast>) -> absy::ImportDirective<'ast> {
-        use crate::absy::NodeValue;
+fn import_directive_to_symbol_vec(
+    import: pest::ImportDirective,
+) -> Vec<absy::SymbolDeclarationNode> {
+    use crate::absy::NodeValue;
 
-        match import {
-            pest::ImportDirective::Main(import) => absy::ImportDirective::Main(
-                imports::Import::new(None, std::path::Path::new(import.source.span.as_str()))
-                    .alias(import.alias.map(|a| a.span.as_str()))
-                    .span(import.span),
-            ),
-            pest::ImportDirective::From(import) => absy::ImportDirective::From(
-                import
-                    .symbols
-                    .iter()
-                    .map(|symbol| {
-                        imports::Import::new(
-                            Some(symbol.symbol.span.as_str()),
-                            std::path::Path::new(import.source.span.as_str()),
-                        )
-                        .alias(
-                            symbol
-                                .alias
-                                .as_ref()
-                                .map(|a| a.span.as_str())
-                                .or_else(|| Some(symbol.symbol.span.as_str())),
-                        )
-                        .span(symbol.span.clone())
-                    })
-                    .collect(),
-            ),
+    match import {
+        pest::ImportDirective::Main(import) => {
+            let span = import.span;
+            let source = Path::new(import.source.span.as_str());
+            let id = "main";
+
+            let import = absy::CanonicalImport {
+                source,
+                id: absy::SymbolIdentifier::from(id).alias(import.alias.map(|a| a.span.as_str())),
```

Comment:
Oh so that's where the optional alias is required. I would suggest moving the alias detection code from the resolver to here, and remove the optional alias.